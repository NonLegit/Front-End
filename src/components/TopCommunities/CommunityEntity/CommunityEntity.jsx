import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import { Box } from '@mui/material';
import { useState, useEffect } from 'react';
import {
  CommunityName, CommunityImage, CommunityIndex, CommunityListItem, Joined, ResposiveBox,
} from './styles';
import numFormatter from '../../../utils/MembersNum';
import PostJoin from '../../SubReddit/PostJoin';

/**
 * represent a row in the community list
 *
 * @component CommunityEntity
 * @property {number} index - rank of the community
 * @property {string} icon - icon of the community
 * @property {string} subredditName - name of the community
 * @property {boolean} status - is the community is growing or not
 * @property {boolean} isJoined - is the user joined in the community
 * @property {boolean} sidebar - is the community inside the right sidebar
 * @property {number} members - number of members in the community
 * @returns {React.Component} CommunityEntity
 */

function CommunityEntity(props) {
  const {
    index, icon, subredditName, status, isJoined, sidebar, members,
  } = props;

  const [joined, setJoined] = useState(isJoined);
  const [showStatus, setShowStatus] = useState(true);

  useEffect(() => {
    if (status) { setShowStatus(status); }
  }, [status]);

  const handleJoin = () => {
    PostJoin(`/subreddits/${subredditName}/subscribe`, joined ? 'unsub' : 'sub');
    setJoined((prev) => !prev);
  };

  const [hover, setHover] = useState(false);

  const handleMouseIn = () => {
    setHover(true);
  };
  const handleMouseOut = () => {
    setHover(false);
  };

  return (
    <CommunityListItem sidebar={sidebar?.toString()}>
      <Box
        sx={{
          display: 'flex',
          alignItems: 'center',
          cursor: 'pointer',
          flex: '1 1 100%',
          flexWrap: 'wrap',
        }}
        onClick={() => { window.open(`${window.location.protocol}/r/${subredditName}`); }}
      >
        <CommunityIndex>
          {index}
        </CommunityIndex>
        {showStatus ? (
          <KeyboardArrowUpIcon
            sx={{ color: '#46d160' }}
          />
        ) : (
          <KeyboardArrowDownIcon
            sx={{ color: '#ea0027' }}
          />
        )}
        <CommunityImage
          sidebar={sidebar?.toString()}
          src={icon}
        />
        <CommunityName variant="caption">
          r/
          {subredditName}
        </CommunityName>

      </Box>
      {!sidebar && (
      <ResposiveBox>
        <Box>
          {!isJoined && (
          <Joined
            variant="contained"
            onClick={handleJoin}
            onMouseEnter={handleMouseIn}
            onMouseLeave={handleMouseOut}
          >
            {(joined ? (hover ? 'Leave' : 'Joined') : 'Join')}
          </Joined>
          )}
        </Box>
        <CommunityIndex sx={{
          display: 'flex', justifyContent: 'flex-end', flexGrow: '1', height: 'fit-content',
        }}
        >
          {numFormatter(members) }
        </CommunityIndex>
      </ResposiveBox>
      )}

    </CommunityListItem>
  );
}

export default CommunityEntity;
