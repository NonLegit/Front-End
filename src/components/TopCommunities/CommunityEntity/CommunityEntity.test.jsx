import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import CommunityEntity from './CommunityEntity';

test('test snapshot', async () => {
  const tree = renderer.create(
    <CommunityEntity />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
