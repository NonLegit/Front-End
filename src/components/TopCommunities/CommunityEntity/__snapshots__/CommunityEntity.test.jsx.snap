// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<li
  className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1l39lt8-MuiListItem-root"
  disabled={false}
>
  <div
    className="MuiBox-root css-l1e50c"
    onClick={[Function]}
  >
    <div
      className="MuiBox-root css-sh1iq1"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-662pq5-MuiSvgIcon-root"
      data-testid="KeyboardArrowUpIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7.41 15.41 12 10.83l4.59 4.58L18 14l-6-6-6 6z"
      />
    </svg>
    <div
      className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1uj8hhm-MuiAvatar-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
        data-testid="PersonIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
    <span
      className="MuiTypography-root MuiTypography-caption css-1yprkys-MuiTypography-root"
    >
      r/
    </span>
  </div>
  <div
    className="MuiBox-root css-1fgbvy9"
  >
    <div
      className="MuiBox-root css-0"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-r2lh7q-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Join
      </button>
    </div>
    <div
      className="MuiBox-root css-gfiief"
    />
  </div>
</li>
`;
